stl/inc/functional

```diff
@@ -1768,31 +1768,57 @@ private:
 
 // FUNCTION TEMPLATE _Build_boyer_moore_delta_2_table
 template <class _RanItPat, class _Pred_eq>
-void _Build_boyer_moore_delta_2_table(_Iter_diff_t<_RanItPat>* _Shifts, const _RanItPat _Pat_first,
-    const _Iter_diff_t<_RanItPat> _Pat_size, _Pred_eq& _Eq) {
-    // builds Boyer-Moore's delta_2 table from a pattern [_Pat_first, _Pat_first + _Pat_size)
-    // pre: _Shifts is a pointer to _Pat_size _Iter_diff_t<_RanIt>s
-    if (_Pat_size == 0) {
+void _Build_boyer_moore_delta_2_table(_Iter_diff_t<_RanItPat>* const _Shifts, const _RanItPat _Pat_first,
+    const _Iter_diff_t<_RanItPat> _Mx, _Pred_eq& _Eq) {
+    // Builds Boyer-Moore's delta_2 table from a pattern [_Pat_first, _Pat_first + _Mx).
+    // pre: _Shifts is a pointer to _Mx _Iter_diff_t<_RanItPat>s.
+    // This is the `dd'` algorithm from "Fast Pattern Matching In Strings" by Knuth, Morris, and Pratt (1977).
+    // Note that the published algorithm used 1-based indexing!
+    // This contains the "Rytter correction" from "Algorithms For String Searching: A Survey" by Baeza-Yates (1989),
+    // originally from "A Correct Preprocessing Algorithm For Boyer-Moore String-Searching" by Rytter (1980).
+
+    if (_Mx == 0) {
         return;
     }
 
     using _Diff = _Iter_diff_t<_RanItPat>;
-    unique_ptr<_Diff[]> _Suffix_fn{::new _Diff[static_cast<size_t>(_Pat_size)]};
-    for (_Diff _Idx = 0; _Idx < _Pat_size; ++_Idx) {
-        _Shifts[_Idx] = 2 * _Pat_size - _Idx - 1;
+
+    const unique_ptr<_Diff[]> _Fx{new _Diff[static_cast<size_t>(_Mx)]};
+
+    for (_Diff _Kx = 1; _Kx <= _Mx; ++_Kx) {
+        _Shifts[_Kx - 1] = 2 * _Mx - _Kx;

```

Comment:
Preexisting: this should probably hoist `2*_Mx` and overflow check it?

```
if ((numeric_limits<_Diff>::max)() - _Mx < _Mx) {
    _Xlength_error("pattern is too long");
}
```

Reply:
Ideally this overflow check shouldn't be needed, I filed #726 to track that.

Reply:
Added overflow check with a slightly rephrased message. I didn't attempt to extract `2 * _Mx` out of the loop (it's a single arithmetic operation that the compiler can easily see).